### PR TITLE
Bluetooth: gatt: Add option to read multiple without variable length

### PR DESCRIPTION
**Pull request by KKopyscinski**
Currently, with EATT enabled, when bt_gatt_read is called with multiple
handles first it'll try to use gatt_read_mult_vl, and if it fails
gatt_read_mult will be used to try again. Add option to skip
the gatt_read_mult_vl and use gatt_read_mult right away. This is needed
by tests that expect BT_ATT_OP_READ_MULT_REQ but support variable
lenght, thus don't return BT_ATT_ERR_NOT_SUPPORTED.

This was affecting:
GATT/CL/GAR/BV-05-C, GATT/CL/GAR/BI-18-C, GATT/CL/GAR/BI-19-C,
GATT/CL/GAR/BI-20-C, GATT/CL/GAR/BI-21-C, GATT/CL/GAR/BI-22-C

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

### CODE REVIEW

File: include/bluetooth/gatt.h [hunk 1]
```diff
@@ -1344,8 +1343,16 @@ struct bt_gatt_read_params {
...
 			uint16_t offset;
 		} single;
-		/** Handles to read in Read Multiple Characteristic Values. */
-		uint16_t *handles;
+		struct {
+			/** Attribute handles to read with Read Multiple
+			 *  Characteristic Values.
+			 */
+			uint16_t *handles;
+			/** If true, we expect that the read characteristics
```
Comment (joerchan) [inline include/bluetooth/gatt.h:1351]: change wording to use the term variable length.

It is actually a bit more complicated.
The READ_MULT operation can be used for values with fixed and known lengths, except for the last value which can be variable.
The READ_MULT_VL operation can be used for values with variable or unknown length.

I think it would be best that we describe this option as more clear on what we will do with this flag.

Suggested wording:
```
/** If true use Read Multiple Variable Length Characteristic Values procedure. 
 *  The values of the set of attributes may be of variable or unknown length.
 *  If false use Read Multiple Characteristic Values procedure.
 *  The values of the set of attributes must be of a known fixed length, with the
 *  exception of the last value that can have a variable length.
 */

File: subsys/bluetooth/shell/gatt.c [hunk 2]
```diff
@@ -340,7 +340,8 @@ static int cmd_mread(const struct shell *shell, size_t argc, char *argv[])
 
 	read_params.func = read_func;
 	read_params.handle_count = i;
-	read_params.handles = h; /* not used in read func */
+	read_params.multi.handles = h; /* not used in read func */
```
Comment (joerchan) [inline subsys/bluetooth/shell/gatt.c:343]: What does the comment mean?

Comment (KKopyscinski) [reply subsys/bluetooth/shell/gatt.c:343]: I'm not sure, as you noticed this commit changes API, and this was obviously not compiling and I had to change it, I left the comment as it was before, as changing this comment is not related to the issue

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -4061,7 +4061,11 @@ int bt_gatt_read(struct bt_conn *conn, struct bt_gatt_read_params *params)
 
 	if (params->handle_count > 1) {
 #if defined(CONFIG_BT_EATT)
```
[the remark this answers is not in the document]
Comment (KKopyscinski) [reply subsys/bluetooth/host/gatt.c:4063]: I'm not sure what you mean? `gatt_read_mult_vl` is compiled only when `CONFIG_BT_EATT` is set, isn't this directive necessary?

File: include/bluetooth/gatt.h [hunk 4]
```diff
@@ -1344,8 +1343,16 @@ struct bt_gatt_read_params {
...
+			 *  have different lengths
+			 */
+			bool variable;
+		} multi;
```
[the remark this answers is not in the document]
Comment (KKopyscinski) [reply include/bluetooth/gatt.h:1355]: So maybe I just should add duplicate of `bt_gatt_read`, named maybe `bt_gatt_read`, that would call just `gatt_read_mult`,  and build it just for testing, so guard it with `#if defined(CONFIG_BT_TESTING)`? That way it wouldn't change the general API

[the remark this answers is not in the document]
Comment (joerchan) [reply include/bluetooth/gatt.h:1355]: Having test-only code wouldn't make much sense. It would make sense to allow the application to use the most efficient read operation needed (related to the ATT_MTU and response length). So I think exposing it at the application level makes sense.
Doing the API change would be the proper solution.

File: include/bluetooth/gatt.h [hunk 5]
```diff
@@ -1332,8 +1332,7 @@ struct bt_gatt_read_params {
 	/** Read attribute callback. */
 	bt_gatt_read_func_t func;
 	/** If equals to 1 single.handle and single.offset are used.
-	 *  If >1 Read Multiple Characteristic Values is performed and handles
-	 *  are used.
+	 *  If >1 multi.handles and multi.len are used.
```
Comment (joerchan) [inline include/bluetooth/gatt.h:1335]: Need to update to multiple here now.

File: subsys/bluetooth/host/gatt.c [hunk 6]
```diff
@@ -4060,11 +4060,11 @@ int bt_gatt_read(struct bt_conn *conn, struct bt_gatt_read_params *params)
...
 
 	if (params->handle_count > 1) {
-#if defined(CONFIG_BT_EATT)
-		return gatt_read_mult_vl(conn, params);
-#else
-		return gatt_read_mult(conn, params);
-#endif /* CONFIG_BT_EATT */
+		if (params->multiple.variable) {
+			return gatt_read_mult_vl(conn, params);
+		} else {
```
Comment (joerchan) [inline subsys/bluetooth/host/gatt.c:4065]: The ifdefs around these should be updated.

```
#else
static int gatt_read_mult(struct bt_conn *conn,
			      struct bt_gatt_read_params *params)
{
	return -ENOTSUP;
}
#endif /* CONFIG_BT_GATT_READ_MULTIPLE */

#if !defined(CONFIG_BT_GATT_READ_MULTIPLE) || !defined(CONFIG_BT_EATT)
static int gatt_read_mult_vl(struct bt_conn *conn,
			     struct bt_gatt_read_params *params)
{
	return -ENOTSUP;
}
#endif
```

**Review by joerchan — CHANGES_REQUESTED** [review]
So we have two options: Either go through the incompatible API change procedure outlined here to move `handles` inside `multiple`.
https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html?highlight=stable%20api#introducing-incompatible-changes

Or we keep `handles` outside of `multiple`, but add `multiple` with the `variable` field (or an anonymous struct?).

Personally I think the API change is the cleanest.

@carlescufi @jhedberg Opinions?

File: include/bluetooth/gatt.h [hunk 7]
```diff
@@ -1332,8 +1332,7 @@ struct bt_gatt_read_params {
 	/** Read attribute callback. */
 	bt_gatt_read_func_t func;
 	/** If equals to 1 single.handle and single.offset are used.
-	 *  If >1 Read Multiple Characteristic Values is performed and handles
-	 *  are used.
+	 *  If >1 multiple.handles are used.
```
Comment (Thalley) [inline include/bluetooth/gatt.h:1335]:
```suggestion
	 *  If greater than 1 multiple.handles are used.
```
or
```suggestion
	 *  If larger than 1 multiple.handles are used.
```

To be consistent with the `equals` on the line before.

**Comment by Thalley** [comment]
@KKopyscinski I'm curious, why are we attempting to even read multiple handles at once in the test case, if we know that the lower tester device doesn't support it? 

> So we have two options: Either go through the incompatible API change procedure outlined here to move `handles` inside `multiple`.
> https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html?highlight=stable%20api#introducing-incompatible-changes
> 
> Or we keep `handles` outside of `multiple`, but add `multiple` with the `variable` field (or an anonymous struct?).
> 
> Personally I think the API change is the cleanest.
> 
> @carlescufi @jhedberg Opinions?

For what it's worth, I agree that the API change is the cleanest.

**Comment by joerchan** [comment]
@KKopyscinski Could you please create the RFC github issue for this incompatible API change?

**Comment by KKopyscinski** [comment]
@joerchan I filed the issue,  should I do something more, like send that email to `devel` mailing list? I cannot add `Stable API Change` label too, could you add them please?

**Comment by joerchan** [comment]
@KKopyscinski Yes, please follow the stable API change procedure.

**Review by jhedberg — APPROVED** [review]

**Comment by galak** [comment]
Please add an update to the release notes for this API change.

**Review by ppryga-nordic — CHANGES_REQUESTED** [review]
I've got a question about lines:
https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3982-L3984

In my humble opinion they are not needed now. Am I right?

If one calls `bt_gatt_read` with `params->multiple.variable == true` then in case of `err == BT_ATT_ERR_NOT_SUPPORTED`:
- Response for `ATT_READ_MULTIPLE_RSP` will be provided. So result will not contain length for variable length characteristics. It may be not possible to interpret the response data.
- https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3985 callback provided by application is executed with  `err = BT_ATT_ERR_NOT_SUPPORTED`.

Is my understanding of the code correct?

**Comment by KKopyscinski** [comment]
@ppryga Maybe someone can confirm this, but as far as I understand it, when device we request read from returns `BT_ATT_ERR_NOT_SUPPORTED` we handle this response as normal, but we also try again reading using `gatt_read_mult`. We expect `BT_ATT_ERR_NOT_SUPPORTED` from devices that support CORE spec <5.2, so not supporting `EATT`. So we test if maybe they support `gatt_read_mult`. This PR doesn't get rid of this fallback method; this PR gives tool to request `gatt_read_mult` straight away, when we know the second device doesn't support `EATT` or maybe we know that characteristic lengths are the same and we want to use more efficient method.

**Comment by ppryga** [comment]
@KKopyscinski I'm not author of the code but... For me, it looks like the fallback implementation that was provided to give possibility to read multiple varying length characteristics without change in API (that is fixed by this PR).
Consider situation described in my last comment: one tries to read #N variable length characteristics. 
In case the other site does not support `EATT` but supports reading of multiple characteristics then he will get response based on `ATT_READ_MULTIPLE_RSP`. This line will be used to call user callback function: https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3941
In case of `ATT_READ_MULTIPLE_VARIABLE_RSP` there will be a number of call to user callback function:
https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3991-L4008

After merge of your PR, the user may expect to receive callback related with `ATT_READ_MULTIPLE_VARIABLE_RSP` only, due to setting `params->multiple.variable == true`. 

I'm not sure it the implicit fallback solution is a right decision unless clearly stated in the API description.
Also If I were the end user, I would like to have a possibility to disable the fallback.

**Comment by Thalley** [comment]
@ppryga @KKopyscinski iirc, the GATT API (GATT, not ATT), does not expose different callbacks from a single or multiple read. The read is handled one by one in any case, so whatever happens at the ATT level does not, and should not, involve the user using the GATT API.

**Comment by Thalley** [comment]
> @Thalley GATT does not expose different callback for different sub-procedures for `bt_gatt_read`.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/include/bluetooth/gatt.h#L1348-L1362
> 
> Handling of responses for `bt_gatt_read` depends on sub-procedure called. In case of this PR `gatt_read_mult_rsp` and `gatt_read_mult_vl_rsp` are responsible for delivery of results of `bt_gatt_read` call to end user.
> 
> `gatt_read_mult_rsp` provides complete PDU received from peer device:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3941
> 
> `gatt_read_mult_vl_rsp` provides response on a per-characterictic basis (loop that iterates over length, value tuples):
> https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3991-L4008
> 
> So handling of responses if different depending on sub-procedures.
> If read-multiple-variable-length-characteristics is expected and user receives complete PDU filled with multiple characteristics values, the PDU will be interpreted wrongly as **single characteristic value**.
> 
> @KKopyscinski
> In my humble opinion, fallback call to `gatt_read_mult` may lead to errors in application level. Thus it should be removed.

I am considering whether an approach where `gatt_read_mult` is handled the same way as `gatt_read` and `gatt_read_mult_vl`, i.e. the callback is called for each entry. The callback handling of `gatt_read_mult` and `gatt_read_mult_vl` has seemingly been implemented by two different people 4 years apart, which probably explains why they were not implemented in the same way. 

The documentation of `bt_gatt_read_params` nor `bt_gatt_read_func_t` really describes the intended behaviour or either multiple reads. Maybe @Vudentz  or @sjanc can weigh in here.

**Comment by ppryga** [comment]
There are two options how to progress with the PR:
- Keep API change as it is in this PR and remove the so called fallback mechanism `gatt_read_mult_rsp` if `gatt_read_mult_vl` fails.
- Introduce more complex change to API following @Thalley concerns:
>I am considering whether an approach where gatt_read_mult is handled the same way as gatt_read and gatt_read_mult_vl, i.e. the callback is called for each entry. The callback handling of gatt_read_mult and gatt_read_mult_vl has seemingly been implemented by two different people 4 years apart, which probably explains why they were not implemented in the same way.
>
>The documentation of bt_gatt_read_params nor bt_gatt_read_func_t really describes the intended behaviour or either multiple reads. Maybe @Vudentz or @sjanc can weigh in here. 

It would be nice to get @Vudentz @joerchan @jhedberg opinion about it. 

Nevertheless, the fallback mechanism should be removed because it may lead to errors on end user site (see: https://github.com/zephyrproject-rtos/zephyr/pull/36281#issuecomment-883117948).

**Review by ppryga-nordic — CHANGES_REQUESTED** [review]
https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3982-L3984

In my humble opinion this should be removed. This kind of a fallback mechanism requires special handling on end user site:
- If variable length read fails expect that there is a fallback.
- How to interpret returned data if variable length characteristics are put into single PDU buffer? How to get actual length of characteristics to parse the returned data?

**Comment by Vudentz** [comment]
> I am considering whether an approach where `gatt_read_mult` is handled the same way as `gatt_read` and `gatt_read_mult_vl`, i.e. the callback is called for each entry. The callback handling of `gatt_read_mult` and `gatt_read_mult_vl` has seemingly been implemented by two different people 4 years apart, which probably explains why they were not implemented in the same way.
> 
> The documentation of `bt_gatt_read_params` nor `bt_gatt_read_func_t` really describes the intended behaviour or either multiple reads. Maybe @Vudentz or @sjanc can weigh in here.

iirc the problem is that for the non-VL procedure we don't know the actual length of each attribute/handle to break them down, so past qualification test this API is pretty useless, also since the client can't really know in advanced if the server would comply with the length it basically unreliable even if we were to add an API for user to enter the expected length, it is just a guess the server could in turn return a different length which would cause unexpected results.

Because of the above I would argue that there is just 2 procedures in GATT that can be used, Read or Read Multiple Variable Length, if the later is not supported the former should be used instead for each handle.

Also if we were to make changes to the API perhaps it would have been better to split them into bt_gatt_read and bt_gatt_read_multiple and either remove the old Read Multiple procedure entirely or make it clear that is unreliable by design and so client has to take action to parse the entire response which would most likely limited to be used for qualification purpose.

**Comment by Vudentz** [comment]
> https://github.com/zephyrproject-rtos/zephyr/blob/2759fc79147467e727c80dbbe3e2fb896361c7fd/subsys/bluetooth/host/gatt.c#L3982-L3984
> 
> In my humble opinion this should be removed. This kind of a fallback mechanism requires special handling on end user site:
> 
> * If variable length read fails expect that there is a fallback.
> * How to interpret returned data if variable length characteristics are put into single PDU buffer? How to get actual length of characteristics to parse the returned data?

Yep, we can't really fallback like that, the non-VL version is unreliable by design even if the client knows the expected length of each handle there is no way to check that the server returns the right length for each attribute read that way, for instance if the server handles changes and there is a mismatch the client won't able to detect that it had happened.

**Comment by Thalley** [comment]
> iirc the problem is that for the non-VL procedure we don't know the actual length of each attribute/handle to break them down, so past qualification test this API is pretty useless, also since the client can't really know in advanced if the server would comply with the length it basically unreliable even if we were to add an API for user to enter the expected length, it is just a guess the server could in turn return a different length which would cause unexpected results.

For most GATT services the characteristics are fixed-lengths and the client will absolutely know this. For those (which is by far a majority of characteristics), it absolutely makes sense to use the read multiple fixed-length request for performance reasons. 

> Also if we were to make changes to the API perhaps it would have been better to split them into bt_gatt_read and bt_gatt_read_multiple and either remove the old Read Multiple procedure entirely or make it clear that is unreliable by design and so client has to take action to parse the entire response which would most likely limited to be used for qualification purpose.

Given the above, where the upper layers (i.e. the specific GATT profiles) determines whether on not it makes sense to use variable or fixed length reads, I think this could make sense. 

So basically 

- if the client knows that 2 or more characteristics are fixed length, it could (should?) use read multiple fixed length
- if the client doesn't know, or knows that 2 or more characteristics are variable length, it could use the read multiple variable length
- else simply read a single characteristic at a time

The current GATT API doesn't support the above choices for the client. The current API is simpler but with more potential overhead, and it begs the question whether or not it really makes sense to modify the API, as @Vudentz also mentioned, to accommodate the additional performance potentially gained by using fixed length reads at all.

**Comment by Vudentz** [comment]
> > iirc the problem is that for the non-VL procedure we don't know the actual length of each attribute/handle to break them down, so past qualification test this API is pretty useless, also since the client can't really know in advanced if the server would comply with the length it basically unreliable even if we were to add an API for user to enter the expected length, it is just a guess the server could in turn return a different length which would cause unexpected results.
> 
> For most GATT services the characteristics are fixed-lengths and the client will absolutely know this. For those (which is by far a majority of characteristics), it absolutely makes sense to use the read multiple fixed-length request for performance reasons.

No unfortunately the client doesn't know it, not without proper caching support which we don't have, also if you are reading a known characteristic UUID you should use Read by Type procedure, not Read Multiple, since the later is prone to handles changes while the former you just give the range where the could be so there is no chance for the server to respond with an unexpected result.

> > Also if we were to make changes to the API perhaps it would have been better to split them into bt_gatt_read and bt_gatt_read_multiple and either remove the old Read Multiple procedure entirely or make it clear that is unreliable by design and so client has to take action to parse the entire response which would most likely limited to be used for qualification purpose.
> 
> Given the above, where the upper layers (i.e. the specific GATT profiles) determines whether on not it makes sense to use variable or fixed length reads, I think this could make sense.
> 
> So basically
> 
> * if the client knows that 2 or more characteristics are fixed length, it could (should?) use read multiple fixed length

I recommended to never use the fixed length given it is unreadable since we can determined if the server would return the expected length, for attributes of the same type Read By Type shall be used.

> * if the client doesn't know, or knows that 2 or more characteristics are variable length, it could use the read multiple variable length
> * else simply read a single characteristic at a time
> 
> The current GATT API doesn't support the above choices for the client. The current API is simpler but with more potential overhead, and it begs the question whether or not it really makes sense to modify the API, as @Vudentz also mentioned, to accommodate the additional performance potentially gained by using fixed length reads at all.

What I was trying to suggest is something along these lines:

bt_gatt_read:
 - Read procedure
 - Read by Type procedure

bt_gatt_read_multiple
  - Read Multiple VL procedure
  - Fallback to bt_gatt_read to read handles one by one if Read Multiple VL is not supported.

The result callback can probably remain the same for both API since these procedures allow the verify proper responses, this should cover what we currently have without bloating the API with a procedure we can't verify the response has been properly formatted.

**Comment by Thalley** [comment]
> No unfortunately the client doesn't know it, not without proper caching support which we don't have, also if you are reading a known characteristic UUID you should use Read by Type procedure, not Read Multiple, since the later is prone to handles changes while the former you just give the range where the could be so there is no chance for the server to respond with an unexpected result.

Why wouldn't the client know it? If the client wants to read e.g. the battery service value (8 bit) from BAS and the body sensor location (8 bit) from HRS, those are two fixed length values that, given that they are defined by the service specification, have lengths known to the client, it will be able to use the read multiple fixed length. Why should you use several read by type procedures if you can do it in a single read multiple fixed length procedure? 

> I recommended to never use the fixed length given it is unreadable since we cannot determined if the server would return the expected length, for attributes of the same type Read By Type shall be used.

It is true that if we read from a non-compliant server, or if we read from the wrong handle, read multiple fixed length may not work, but neither would any other read.

**Comment by sjanc** [comment]
I think it should be up to user to decide which procedure to use and thus I'm fine with current version of this PR (ie no fallback mechanism).

If at some point host can cache remote features and fail early to avoid extra roundtrip even better, but that will not affect on how application is using the api.

**Review by sjanc — APPROVED** [review]

**Comment by KKopyscinski** [comment]
Removed the fallback to read multiple

**Review by Thalley — APPROVED** [review]

**Review by joerchan — APPROVED** [review]

**Review by ppryga-nordic — APPROVED** [review]
